ResearchSuite/ResearchSuite/RSDMotionRecorder.swift

```diff
@@ -0,0 +1,585 @@
+        self.x = data.vector.x
+        self.y = data.vector.y
+        self.z = data.vector.z
+    }
+    
+    /// Initialize from a `CMDeviceMotion` for a given sensor type or calculated vector.
+    /// - parameters:
+    ///     - startUptime: System clock uptime when the recorder was started.
+    ///     - stepPath: The current step path.
+    ///     - data: The `CMDeviceMotion` data sample from which to record information.
+    ///     - referenceFrame: The `CMAttitudeReferenceFrame` for this recording.
+    ///     - sensorType: The recorder type for which to record the vector.
+    public init?(startUptime: TimeInterval, stepPath: String, data: CMDeviceMotion, referenceFrame: CMAttitudeReferenceFrame, sensorType: RSDMotionRecorderType) {
+        
+        var eventAccuracy: Int?
+        var referenceCoordinate: RSDAttitudeReferenceFrame?
+        let vector: RSDVector
+        var w: Double?
+        var heading: Double?
+        
+        switch sensorType {
+        case .attitude:
+            vector = data.attitude.quaternion
+            w = data.attitude.quaternion.w
+            referenceCoordinate = RSDAttitudeReferenceFrame(frame: referenceFrame)
+            eventAccuracy = Int(data.magneticField.accuracy.rawValue)
+            if #available(iOS 11.0, *) {
+                heading = (data.heading >= 0) ? data.heading : nil
+            }
+            
+        case .gravity:
+            vector = data.gravity
+            
+        case .magneticField:
+            vector = data.magneticField.field
+            eventAccuracy = Int(data.magneticField.accuracy.rawValue)
+            if #available(iOS 11.0, *) {
+                heading = data.heading
+            }
+            
+        case .rotationRate:
+            vector = data.rotationRate
+            
+        case .userAcceleration:
+            vector = data.userAcceleration
+            
+        default:
+            return nil
+        }
+        
+        self.uptime = data.timestamp
+        self.timestamp = data.timestamp - startUptime
+        self.stepPath = stepPath
+        self.timestampDate = nil
+        self.sensorType = sensorType
+        self.eventAccuracy = eventAccuracy
+        self.referenceCoordinate = referenceCoordinate
+        self.x = vector.x
+        self.y = vector.y
+        self.z = vector.z
+        self.w = w
+        self.heading = heading
+    }
+}
+
+/// A string-value representation for the attitude reference frame.
+public enum RSDAttitudeReferenceFrame : String, Codable {
+    
+    /// Describes a reference frame in which the Z axis is vertical and the X axis points in
+    /// an arbitrary direction in the horizontal plane.
+    case xArbitraryZVertical = "Z-Up"
+    
+    /// Describes a reference frame in which the Z axis is vertical and the X axis points toward
+    /// magnetic north.
+    ///
+    /// - note: Using this reference frame may require device movement to calibrate the magnetometer.
```

Comment:
should it say "user interaction" instead of "device movement" just to be clear why this is a potential problem? or something along those lines?